source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -64,12 +65,46 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
   return FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
+FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  if (processing_complete_) {
+    return FilterDataStatus::Continue;
+  }
+
+  switch (processing_mode_.request_body_mode()) {
+  case ProcessingMode::BUFFERED:
+  case ProcessingMode::BUFFERED_PARTIAL:
+    ENVOY_LOG(warn, "Ignoring unimplemented request body processing mode BUFFERED");
+    return FilterDataStatus::Continue;
+  case ProcessingMode::STREAMED:
+    return decodeDataStreamed(data, end_stream);
+  case ProcessingMode::NONE:
+  default:
+    return FilterDataStatus::Continue;
+  }
+}
+
+FilterDataStatus Filter::decodeDataStreamed(Buffer::Instance& data, bool end_stream) {
+  ENVOY_BUG(request_state_ == FilterState::IDLE, "Filter request state should be idle");
+  openStream();
+  request_body_chunk_ = &data;
+  ProcessingRequest req;
+  auto* body_req = req.mutable_request_body();
+  body_req->set_end_of_stream(end_stream);
+  body_req->set_body(data.toString());
```

Comment:
I think you need to keep how many bytes sent to the ext proc, or you'll send same buffer multiple times. StopIterationAndWatermark won't clear the buffer IIRC.

Reply:
In this case of the "streaming" body mode, we want to send each individual chunk of the body to the ext_proc server and let it respond if it wants to change it. So I think this is right. However I can't prove that so I am going to add some additional integration tests (which are hard BTW because I'm not totally sure how to guarantee that the integration tests do what I want). 

Later on I want to add a "buffered" body mode which will send the whole thing at once.

Reply:
Also -- I hate having to save a pointer to the buffer here and assume that it's valid later. Same with the headers. It works but I worry about the filter semantics changing someday. If you have any other ideas I'd appreciate.

Reply:
What I'm saying is that because you don't `Continue` here but `StopIterationAndWatermark`, if you received chuncks like `a`, `b`, `c`, you are going to send `ProcessRequest('a')`, `ProcessRequest('ab')`, `ProcessRequest('abc')`, because the buffer won't be drained. I don't think this is what you want?

Reply:
You're right -- both "Buffer" and "Watermark" buffer. I think I want StopIterationNoBuffer in this particular case, but now I see that it pipelines callbacks -- (en|de)codeData get called before continue(En|De)coding, and that messes up the state machine. So, I need to go back to the drawing board on this.

Reply:
@gbrail also you might find https://github.com/envoyproxy/envoy/blob/main/source/docs/flow_control.md#http1-and-http2-filters useful (and the whole doc there) if you haven't seen it before.